File: graph_theory/dijkstra_shortest_path/dijkstra.cc
#include <iostream>
#include <vector>
#include <queue>
#include <iterator>
#include <algorithm>
#include "utils/graph.h"

void dijkstra(const Graph &g, int s, std::vector<Weight> &dist, std::vector<int> &prev)
{
    const int n = g.size();
    dist.assign(n, INF); dist[s] = 0;
    prev.assign(n, -1);

    std::priority_queue<Edge> Q;
    for (Q.push(Edge(-2, s, 0)); !Q.empty(); ) {
        Edge e = Q.top(); Q.pop();
        if (prev[e.dst] != -1) continue;

        prev[e.dst] = e.src;
        FOR(f, g[e.dst]) if (dist[f->dst] > e.weight + f->weight) {
            dist[f->dst] = e.weight + f->weight;
            Q.push(Edge(f->src, f->dst, e.weight + f->weight));
        }
    }
}

std::vector<int> buildPath(const std::vector<int> &prev, int t)
{
    std::vector<int> path;
    for (int u = t; u >= 0; u = prev[u])
        path.push_back(u);
    reverse(path.begin(), path.end());
    return path;
}

int main(void)
{
    std::vector<Weight> dist;
    std::vector<int> prev;
    Graph g;

    g.resize(3);
    g[0].push_back(Edge(0, 1, 101));
    g[1].push_back(Edge(1, 2, 1));
    g[2].push_back(Edge(2, 0, 10000));

    dijkstra(g, 0, dist, prev);
    std::vector<int> path = buildPath(prev, 2);

    std::cout << dist[2] << std::endl;
    std::copy(path.begin(), path.end(), std::ostream_iterator<int>(std::cout, " "));

    return 0;
}

File: graph_theory/shortest_path/dijkstra/dijkstra.py
import queue

INF = 1000000000

class Edge():
    def __init__(self, src, dst, weight):
        self.src = src
        self.dst = dst
        self.weight = weight

    def __lt__(self, other):
        if self.weight != other.weight:
            return self.weight < other.weight
        elif self.dst != other.dst:
            return self.dst < other.dst
        else:
            return self.src < other.src


def dijkstra(g, s):
    n = len(g)
    dist = [INF] * n
    dist[s] = 0

    prev = [-1] * n

    Q = queue.PriorityQueue()
    Q.put(Edge(-2, s, 0))
    while not Q.empty():
        e = Q.get()

        if prev[e.dst] != -1:
            continue

        prev[e.dst] = e.src
        for f in g[e.dst]:
            if dist[f.dst] > e.weight + f.weight:
                dist[f.dst] = e.weight + f.weight
                Q.put(Edge(f.src, f.dst, e.weight + f.weight))

    return (dist, prev)


if __name__ == "__main__":
    n, m = map(int, input().split())
    g = [[]] * n
    for i in range(0, m):
        a1, a2, w = map(int, input().split())
        g[a1].append(Edge(a1, a2, w))

    dist, prev = dijkstra(g, 0)
    print(dist)
